File: OJ/usaco/cpp/inflate.cpp
/*
ID:oifox20071
LANG:C++
PROG:inflate
*/
#include <iostream>
#include <cstring>
using namespace std;
const int maxm = 10001;
const int maxn = 10000;
int f[maxm], p[maxn], v[maxn];
int main()
{
	freopen("inflate.in", "r", stdin); freopen("inflate.out", "w", stdout);
	int n, m;
	cin >> m >> n;
	for (int i = 0; i < n; ++ i)
		cin >> p[i] >> v[i];
	memset(f, 0, sizeof(f));
	for (int i = 0; i < n; ++ i)
		for (int j = v[i]; j <= m; ++ j)
			f[j] = max(f[j], f[j - v[i]] + p[i]);
	cout << f[m] << endl;
	return 0;
}

File: OJ/usaco/pascal/s3/s31/inflate.pas
{
ID:lijianx1
PROG:inflate
LANG:PASCAL
}
program inflate;
const
	maxn = 10000;
	maxm = 10000;
var
	f:array[0..maxm] of longint;
	v,p:array[1..maxn] of longint;
	i,j,m,n:longint;
	
begin
	assign(input,'inflate.in'); reset(input);
	readln(m,n);
	for i:=1 to n do readln(v[i],p[i]);
	close(input);
	fillchar(f,sizeof(f),0);
	for i:=1 to n do
		for j:=p[i] to m do
			if (f[j - p[i]] + v[i] > f[j]) then f[j]:=f[j - p[i]] + v[i];
	assign(output,'inflate.out'); rewrite(output);
	writeln(f[m]);
	close(output);
end.